src/main/java/com/cloudant/search3/Search.java

```diff
@@ -277,99 +300,46 @@ public void close() {
         } catch (final InterruptedException e) {
             // Ignored.
         }
-        handlers.forEach((index, handler) -> {
-            try {
-                handler.close();
-            } catch (final IOException e) {
-                LOGGER.catching(e);
-            }
-        });
-        handlers.clear();
+        handlers.invalidateAll();
     }
 
-    private <T> void execute(
+    private <R> R execute(
             final Index index,
-            final StreamObserver<T> responseObserver,
-            final LuceneConsumer<SearchHandler> f) {
+            final LuceneFunction<SearchHandler, R> f) throws IOException, ParseException, ExecutionException {
         try {
-            final SearchHandler handler = getOrOpen(index);
-            f.accept(handler);
+            final SearchHandler handler = handlers.get(new SearchCacheKey(index));
+            return f.apply(handler);
+        } catch (ExecutionException e) {
+            failedHandler(index, e);
+            throw e;
         } catch (final IOException | AlreadyClosedException e) {
```

Comment:
Can we just list these exception types in the previous catch as well? The are handled identically.

Reply:
this is outside of the PR I think? (diff tab won't navigate me to this section of code). But, yes, I think we could collapse these, this block has evolved over time. It was important to choose when to call failedHandler versus not.